dir.c

```diff
@@ -810,10 +828,22 @@ fundamental_encoding_p(rb_encoding *enc)
 # define READDIR(dir, enc) rb_w32_readdir((dir), (enc))
 # define READDIR_NOGVL READDIR
 #else
+NORETURN(static void *sys_failure(void *function));
+static void *
+sys_failure(void *function)
+{
+    rb_sys_fail(function);
+}
+
 static void *
 nogvl_readdir(void *dir)
 {
-    return readdir(dir);
+    errno = 0;
+    if ((dir = readdir(dir)) == NULL) {
+        if (errno)
+            rb_thread_call_with_gvl(sys_failure, (void *)"readdir");
+    }
```

Comment:
This looks to make the check in `dir_read` unneeded.

And maybe this needs `rb_errno_set` and `rb_errno` for M:N thread scheduler.

```suggestion
    rb_errno_set(0);
    if ((dir = readdir(dir)) == NULL) {
        if (rb_errno())
            rb_thread_call_with_gvl(sys_failure, (void *)"readdir");
    }
```